openhtf/exe/__init__.py

```diff
@@ -122,40 +131,113 @@ def _RunTestLoop(self):
       # Top level steps required to run a single iteration of the Test.
       _LOG.info('Starting test %s', self.test.code_info.name)
 
-      # Save exit_stack on self so we can access it when Stop() is called.
-      self._current_exit_stack = exit_stack
-      exit_stack.callback(lambda: setattr(self, '_current_exit_stack', None))
-
-      plug_manager = self._MakePlugManager(suppressor)
-      executor = self._MakePhaseExecutor()
-
+      # Any access to self._exit_stack must be done while holding this lock.
+      with self._lock:
+        # Save exit_stack on self so we can access it if Stop() is called.
+        self._exit_stack = exit_stack
+
+        # Reset some stateful attributes to None when we're done with all other
+        # teardown.
+        exit_stack.callback(lambda: setattr(self, '_exit_stack', None))
+        exit_stack.callback(lambda: setattr(self, '_test_state', None))
+        exit_stack.callback(lambda: setattr(self, '_output_thread', None))
+
+      # Wait here until the test start trigger returns a DUT ID.  Don't hold
+      # self._lock while we do this, or else calls to Stop() will deadlock.
+      dut_id = self._WaitForTestStart(suppressor)
+
+      with self._lock:
+        if not self._exit_stack:
+          # We shouldn't get here, but just in case something went weird with a
+          # call to Stop() and we ended up resuming execution here but the
+          # exit stack was already cleared, bail.
+          raise TestStopError('Test Stopped.')
+
+        # After teardown of everything else, wait on the output thread before
+        # attempting to start another test.
+        exit_stack.callback(self._MaybeJoinOutputThread)
```

Comment:
I would only add this to the exit stack after the output thread has been created